File: android/src/main/java/dev/robingenz/capacitor/screenorientation/ScreenOrientationPlugin.java
package dev.robingenz.capacitor.screenorientation;

import android.content.res.Configuration;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "ScreenOrientation")
public class ScreenOrientationPlugin extends Plugin {

    public static final String SCREEN_ORIENTATION_CHANGE_EVENT = "screenOrientationChange";

    private ScreenOrientation implementation;

    @Override
    public void load() {
        implementation = new ScreenOrientation(getBridge());
        implementation.setOrientationChangeListener(this::updateOrientation);
    }

    @Override
    public void handleOnConfigurationChanged(Configuration newConfig) {
        super.handleOnConfigurationChanged(newConfig);
        implementation.handleOnConfigurationChanged(newConfig);
    }

    @PluginMethod
    public void lock(PluginCall call) {
        String orientationType = call.getString("type");
        implementation.lock(orientationType);
        call.resolve();
    }

    @PluginMethod
    public void unlock(PluginCall call) {
        implementation.unlock();
        call.resolve();
    }

    @PluginMethod
    public void getCurrentOrientation(PluginCall call) {
        JSObject ret = new JSObject();
        String orientationType = implementation.getCurrentOrientationType();
        ret.put("type", orientationType);
        call.resolve(ret);
    }

    private void updateOrientation() {
        JSObject ret = new JSObject();
        String orientationType = implementation.getCurrentOrientationType();
        ret.put("type", orientationType);
        notifyListeners(SCREEN_ORIENTATION_CHANGE_EVENT, ret);
    }
}

File: ios/Plugin/ScreenOrientationPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ScreenOrientationPlugin)
public class ScreenOrientationPlugin: CAPPlugin {
    public let screenOrientationChangeEvent = "screenOrientationChange"
    private var implementation: ScreenOrientation?

    override public func load() {
        self.implementation = ScreenOrientation(plugin: self)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    @objc func lock(_ call: CAPPluginCall) {
        let orientationType = call.getString("type") ?? ""
        implementation?.lock(orientationType, completion: {
            call.resolve()
        })
    }

    @objc func unlock(_ call: CAPPluginCall) {
        implementation?.unlock(completion: {
            call.resolve()
        })
    }

    @objc func getCurrentOrientation(_ call: CAPPluginCall) {
        implementation?.getCurrentOrientationType(completion: { orientationType in
            call.resolve([
                "type": orientationType
            ])
        })
    }

    @objc func notifyOrientationChangeListeners(_ orientationType: String) {
        self.notifyListeners(self.screenOrientationChangeEvent, data: [
            "type": orientationType
        ])
    }
}
